### ISSUE TITLE
Separate `poliastro.core` from the rest?

### ISSUE THREAD

**Issue by astrojuanlu** [poliastro/poliastro#1207]
In today's community call, @jorgepiloto and I discussed the possibility to separate `poliastro.core` from the rest. This is somewhat related to #168, but uses a different method: by recognizing that `poliastro.core` is interesting and useful on its own, we split it off as a separate project, and make poliastro depend on it.

Luckily, `poliastro.core` only depends on NumPy and numba:

```
$ grep import src/poliastro/core/ -rh --exclude=*.pyc | grep -v '>' | sort | uniq
from ..angles import (
from .angles import E_to_nu, F_to_nu
from astropy import units as u
from ..elements import coe2rv, rv2coe
from .farnocchia import farnocchia, farnocchia_coe
from math import gamma
from numba import njit as jit
from numba import njit as jit, prange
from numpy import cos, cross, sin, sqrt
from numpy import cos, sin
from numpy import cross
from numpy import cross, pi
from numpy.linalg import norm
from poliastro.core.elements import coe_rotation_matrix, rv2coe, rv_pqw
from poliastro.core.hyper import hyp2f1b
from poliastro.core.stumpff import c2, c3
from poliastro.core.util import circular_velocity
from ..stumpff import c2, c3
from .util import rotation_matrix
import numpy as np
```

Advantages:

- Easier to contribute "raw algorithms" (see also our recent efforts in #1189)
- Users that just want the core functionality can just use the new project
- Users that prefer alternatives to `astropy.units` can bring their own (for example https://pypi.org/project/Pint/ and possibly others)

All of this is possible by installing poliastro and just ignoring the safe API, but the disadvantage is that it requires a lot of heavy dependencies (SciPy, Astropy, Plotly, matplotlib, others).

Disadvantages:

- More overhead
- We need coordination between the two projects

Thoughts?

**Comment by Yash-10** [comment]
Hello @astrojuanlu,

I was thinking of some points, and it would be great to get everyone's feedback on the below:

a. Currently, there seem to be some inconsistencies in how `core` methods expect inputs (for example, some methods expect `k` to be in `u.m **3 / u.s ** 2` while in `u.km **3 / u.s ** 2` in other methods). If we attempt standardise the inputs and assert that they must have the same particular units throughout `poliastro.core`, it would be great! As a result, factoring `core` into a separate project would really seem to be a good choice and make it more convenient to implement.

b. Since some methods currently cannot be implemented in `core` because they don't use raw orbital algorithms directly but leverage high-level API's (for example, the `perturbations` module), there might be two options:

1. Finding a way to re-implement all the methods (which use third-party libraries in some way) efficiently (in terms of computation speed and memory). This would involve a great deal of validation and bench-marking between the third-party implementations and poliastro's executions, thus making the process cumbersome. Or
2. Declare functionalities that cannot be implemented in `core` as "impossible" to be present in `core` and direct users to non-core poliastro for using them.

Thanks!

**Comment by astrojuanlu** [comment]
Thanks @Yash-10 !

a. It's a very good point. We should tackle #707 at some point to remove this inconsistency.
b. You're right. In this case, I'd prefer keeping them in `core`, but without being accelerated by numba.

**Comment by Yash-10** [comment]
> We should tackle #707 at some point to remove this inconsistency.

Yes! I shall take a closer look at #707 to understand more about it.

</br>

> In this case, I'd prefer keeping them in core, but without being accelerated by numba.

Thanks for your confirmation!

**Comment by astrojuanlu** [comment]
Very personal comment ahead:

I'd like to give this more priority. I voiced some thoughts on [this Twitter thread](https://twitter.com/juanluisback/status/1436242830834511874):

> @/choldgraf: Is there a good write up about the pros/cons of an open source project following best practices in software design at the expense of being less accessible to newcomers, vs intentionally using more straightforward logic, tooling, etc in order to make it easier to contribute?
> @/juanluisback: This has affected me directly: my main open-source project, @/poliastro_py, suffers from slow development because I don't feel I am able to "get the API right". And as a result, it affects all its contributors. I barely started working on addressing this.
> @/juanluisback: For me it's a mix of: (1) knowledge gaps (I don't know Computer Science, SOLID principles, Design Patterns), (2) trying to get it *perfect* and not knowing when to stop, and (3) I actually enjoy writing algorithms more than designing APIs when things get hairy.

I feel that API design conversations are essential for a high-quality, reusable software library. But at the same time, I feel less motivated than ever to have them. [The project wiki](https://github.com/poliastro/poliastro/wiki/) is full of my past ideas, and I have several notebooks at home with handwritten notes, but in the end, I never feel confident enough about decisions. Things were easy when we only had `Orbit` and `Maneuver`, but now it's much more complicated, and it's will get trickier with time.

On the other hand, I have so much fun reading through ancient papers and implementing algorithms. It's ungrateful and difficult, but exciting still.

As a result, I think I would be much more motivated to keep contributing to the poliastro ecosystem if I could stick to the things I enjoy the most:

1. Community building, documentation, developer experience
2. Computational physics, open science

and if I could leave API design for more motivated folks that use the library on a daily basis.

In summary, I don't think I'm going to tackle this in the immediate future, but I'm putting it high on my priorities list, and I'll try to have it ready for next year's GSOC.

**Comment by MLopez-Ibanez** [comment]
My two cents:

* Physically separating into different packages within github seems an unnecessary overhead. If separate packages are desired, it would be better to be able to build separate packages from a common source tree.
* Before separating, one should come up with "border" rules that could hopefully be checked by the testers. For example, one rule that is already violated is never use astropy inside core (in particular units). 
```
core/perturbations.py:from astropy import units as u
```
* Another border rule should be a list of "standard" units for all of core. Those units should be chosen to simplify code. For example, always use radians. This is also not enforced at the moment. The documentation of all core functions should describe those units to force people adding/updating functions to think about the units. A tester script can then check that the documentation satisfies the standard units by checking for strings such as "(units: km)" "(units: rad)" "(units: seconds)". If something unrecognized is found like "(units: deg)", reject it.

**Comment by astrojuanlu** [comment]
Thanks for chiming in @MLopez-Ibanez ! I agree that maybe we don't need 2 repositories for 2 projects, I hadn't considered that before. Keeping them together avoids losing history, git context, etc.

Regarding your other two points, they are tracked in #1256 and #707. The former is much more urgent than the latter, and we could enforce it by using import-linter as I wrote in https://github.com/poliastro/poliastro/issues/1256#issuecomment-864532393

**Comment by s-m-e** [comment]
Just some homework and an overview of all the places where `poliastro` currently, as of d85c9bd1e2e8c19765d34e6ee3d672cb51da22cc, talks to its core.

https://github.com/pleiszenburg/bulk_propagate/blob/3d132ace90cdb28252620fc2b5eee5cfed42cb20/core_interface_01.ipynb

**Comment by astrojuanlu** [comment]
Thanks @s-m-e. I already see lots of things there I don't like 😂 some cleanup is warranted for sure...

**Comment by astrojuanlu** [comment]
We're _not_ doing this before 1.0.

**Comment by jorgepiloto** [comment]
I have been thinking a lot about this issue and thought about turning poliastro into a metapackage and creating new repositories:

```console
- poliastroy/poliastro
- poliastro/core
- poliastro/plotting
- poliastro/api
```

All packages developed for poliastro should follow the same namespace, leading to the following result:

```python
from poliastro.core import __version__
from poliastro.plotting import __version__
from poliastro.bodies import ...
...
```

Previous structure would allow to install the desired component of poliastro:

```console
python -m pip install poliastro  # Installs the whole poliastro ecosystem. Only for brave people.
python -m pip install poliastro[core]  # Only installs fundamental astro-routines
python -m pip install poliastro[api]  # A more readable interface for working with core
python -m pip install poliastro[plotting]  # Install all heavy plotting libraries
```

Previous Python packages can be contained into different repositories of the organization or following a mono-repo approach. I just investigated this feature and it is really useful. The main advantage is that the whole CI/CD takes place into a single repository and guarantees integration between the different libraries.
